compiler/rustc_driver_impl/src/lib.rs

```diff
@@ -1237,6 +1238,26 @@ pub fn handle_options(early_dcx: &EarlyDiagCtxt, args: &[String]) -> Option<geto
         return None;
     }
 
+    // To avoid confusion, emit warning if no space
+    // between `-o` and arg, e.g.`-optimize`, `-out-dir`, is applied, see issue #142812
+    if let Some(name) = matches.opt_str("o")
+        && let Some(suspect) = args.iter().find(|arg| arg.starts_with("-o") && *arg != "-o")
+    {
+        let confusables = ["optimize", "o0", "o1", "o2", "o3", "ofast", "og", "os", "oz"];
+        if let Some(confusable) = check_confusables(&suspect, &confusables) {
+            early_dcx.early_warn(
+                "option `-o` has no space between flag name and value, which can be confusing",
+            );
+            early_dcx.early_note(format!(
+                "option `-o {}` is applied instead of a flag named `o{}` to specify output filename `{}`",
+                name, name, name
+            ));
```

Comment:
Nit:

```suggestion
            early_dcx.early_note(format!(
                "output filename `-o {name}` is applied instead of a flag named `o{name}`"
            ));
```